### PR TITLE
fix(test): Show full Zebra test panic details in CI logs

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

We're seeing some Zebra panics in CI. Our custom test harness just reports "The application panicked (crashed).", with no details.

Example logs:
https://github.com/ZcashFoundation/zebra/runs/7981879420?check_suite_focus=true#step:6:1103

Depends-On: #4945

## Solution

- Add output logs to test context, and add test harness tests to check it works
- Let empty test child logs be read again (and produce empty output

Related refactors:
- Add an `ignore_exited` argument to test child process `kill()`
- Handle test failure regexes using Result::Err, rather than panicking

Related changes:
- Log the command timeout when an acceptance test fails

## Review

This is an urgent fix, so we can see panic details in our CI logs.

### Reviewer Checklist

  - [x] CI passes

### CODE REVIEW

File: zebra-test/src/command.rs [hunk 1]
```diff
@@ -526,12 +553,12 @@ impl<T> TestChild<T> {
     {
         self.apply_failure_regexes_to_outputs();
 
-        if let Some(Ok(line)) = self.stdout.as_mut().and_then(|iter| iter.next()) {
+        if let Some(line) = self.stdout.as_mut().and_then(|iter| iter.next()) {
```
Comment (arya2) [inline zebra-test/src/command.rs]: small nit: `line_result` or `line_res` might be more descriptive than `line` here and on line 602

**Comment by teor2345** [comment]
Now we know what the panic bug is, this isn't a critical priority any more.

**Review by arya2 — DISMISSED** [review]

File: zebra-test/src/command.rs [hunk 2]
```diff
@@ -771,16 +834,15 @@ impl<T> TestChild<T> {
 
         if self.is_running() {
             // If the process exits between is_running and kill, we will see
-            // spurious errors here. If that happens, ignore "no such process"
+            // spurious errors here. So we want to ignore "no such process"
             // errors from kill.
-            self.kill()?;
+            self.kill(true).context_from(self)?;
```
Comment (teor2345) [inline zebra-test/src/command.rs]: Revert this - we're getting the context higher up

File: zebrad/tests/common/sync.rs [hunk 3]
```diff
@@ -248,7 +248,7 @@ pub fn sync_until(
 
         // make sure the child process is dead
         // if it has already exited, ignore that error
-        let _ = child.kill();
+        let _ = child.kill(true);
```
Comment (teor2345) [inline zebrad/tests/common/sync.rs]: Stop ignoring other kill errors here.

File: zebra-test/src/command.rs [hunk 4]
```diff
@@ -3,7 +3,7 @@
 use std::{
```
Comment (teor2345) [inline zebra-test/src/command.rs:3]: Rebase on PR #4945 before pushing

File: zebra-test/src/command.rs (hunk 1 above)
Comment (teor2345) [reply zebra-test/src/command.rs]: Fixed in commit 6c0ec540d7

File: zebra-test/src/command.rs (hunk 2 above)
Comment (teor2345) [reply zebra-test/src/command.rs]: Fixed in 9074205b29

File: zebrad/tests/common/sync.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/tests/common/sync.rs]: Fixed in 92afab915e

File: zebra-test/src/command.rs (hunk 4 above)
Comment (teor2345) [reply zebra-test/src/command.rs:3]: Done!

**Comment by teor2345** [comment]
Marking this as critical because we're still seeing panics, and we need to know what they are. So we want to merge this first.

**Comment by teor2345** [comment]
@mergifyio update

**Comment by teor2345** [comment]
I added some extra test debug output in commit 9d4c709d71. Alfredo and I talked about it earlier this week to make it easier to diagnose test failures.

I was going to put it in PR #4958, but that might have conflicted with this PR, so I put it here instead.

**Comment by teor2345** [comment]
@mergifyio update

**Review by gustavovalverde — APPROVED** [review]
Approving as this is critical and CI is passing.

**Pull request merged**